Python/optimizer_analysis.c

```diff
@@ -9,10 +20,457 @@
+        return NULL;
+    }
+    sym_set_flag(res, NOT_NULL);
+    return res;
+}
+
+static inline _Py_UOpsSymType*
+sym_new_known_pytype(_Py_UOpsAbstractInterpContext *ctx,
+                      PyTypeObject *typ)
+{
+    _Py_UOpsSymType *res = _Py_UOpsSymType_New(ctx,NULL);
+    if (res == NULL) {
+        return NULL;
+    }
+    sym_set_pytype(res, typ);
+    return res;
+}
+
+// Takes a borrowed reference to const_val.
+static inline _Py_UOpsSymType*
+sym_new_const(_Py_UOpsAbstractInterpContext *ctx, PyObject *const_val)
+{
+    assert(const_val != NULL);
+    _Py_UOpsSymType *temp = _Py_UOpsSymType_New(
+        ctx,
+        const_val
+    );
+    if (temp == NULL) {
+        return NULL;
+    }
+    sym_set_pytype(temp, Py_TYPE(const_val));
+    sym_set_flag(temp, TRUE_CONST);
+    return temp;
+}
+
+static _Py_UOpsSymType*
+sym_new_null(_Py_UOpsAbstractInterpContext *ctx)
+{
+    if (ctx->frequent_syms.push_nulL_sym != NULL) {
+        return ctx->frequent_syms.push_nulL_sym;
+    }
+    _Py_UOpsSymType *null_sym = sym_new_unknown(ctx);
+    if (null_sym == NULL) {
+        return NULL;
+    }
+    sym_set_pytype(null_sym, NULL);
+    ctx->frequent_syms.push_nulL_sym = null_sym;
+    return null_sym;
+}
+
+
+static inline bool
+sym_matches_pytype(_Py_UOpsSymType *sym, PyTypeObject *typ)
+{
+    assert(typ == NULL || PyType_Check(typ));
+    if (!sym_has_flag(sym, KNOWN_TYPE)) {
+        return false;
+    }
+    return sym->typ == typ;
+}
+
+
+static inline bool
+op_is_end(uint32_t opcode)
+{
+    return opcode == _EXIT_TRACE || opcode == _JUMP_TO_TOP;
+}
+
+
+static inline bool
+op_is_bookkeeping(uint32_t opcode) {
+    return (opcode == _SET_IP ||
+            opcode == _CHECK_VALIDITY ||
+            opcode == _SAVE_RETURN_OFFSET ||
+            opcode == _RESUME_CHECK);
+}
+
+static inline bool
+op_is_data_movement_only(uint32_t opcode) {
+    return (opcode == _STORE_FAST ||
+        opcode == STORE_FAST_MAYBE_NULL ||
+        opcode == _PUSH_FRAME ||
+        opcode == _POP_FRAME);
+}
+
+
+static inline int
+emit_i(uops_emitter *emitter,
+       _PyUOpInstruction inst)
+{
+    if (emitter->writebuffer + emitter->curr_i >= emitter->writebuffer_end) {
+        OPT_STAT_INC(optimizer_failure_reason_no_writebuffer);
+        DPRINTF(1, "out of emission space\n");
+        return -1;
+    }
+    if (inst.opcode == _NOP) {
+        return 0;
+    }
+    DPRINTF(2, "Emitting instruction at [%d] op: %s, oparg: %d, target: %d, operand: %" PRIu64 " \n",
+            emitter->curr_i,
+            _PyOpcode_uop_name[inst.opcode],
+            inst.oparg,
+            inst.target,
+            inst.operand);
+    emitter->writebuffer[emitter->curr_i] = inst;
+    emitter->curr_i++;
+    return 0;
+}
+
+static int
+uop_abstract_interpret_single_inst(
+    _PyUOpInstruction *inst,
+    _PyUOpInstruction *end,
+    _Py_UOpsAbstractInterpContext *ctx
+)
+{
+#define STACK_LEVEL()     ((int)(stack_pointer - ctx->frame->stack))
```

Comment:
Can you define macros outside functions? It makes the function hard to follow. I didn't realize that the code below was in a function due to the macros.